### ISSUE TITLE
 some reports from users complaining about "native thread exited without detaching..."

### ISSUE THREAD

**Issue by akshayaurora** [kivy/pyjnius#107]
http://yltechblog.blogspot.in/2013/05/jni-detaching-threads-native-thread.html <- seems like a solution

**Comment by brentpicasso** [comment]
Still seeing this problem. Being able to use pyjnius from a thread is a critical, fundamental requirement and without it makes this framework fairly useless for serious applications.  Is there someone we can throw money at to help fix it?

In my case I am using pyjnius to talk to Bluetooth on android, running in a background service, passing messages to the main app using OSC.

I've tried a variety of approaches to prevent the background task from crashing - joining the thread before the service exits, or setting the python thread as a daemon and exiting the service. Both result in a segfault and the message "Unfortunately <your app> has stopped". 

At this point I would be happy letting it segfault on exit if I could suppress that message!

Current version of the background service:
https://github.com/autosportlabs/RaceCapture_App/blob/master/service/main.py

Thoughts? Should I just go to a native java service or use the NDK?

**Comment by tito** [comment]
Just for sharing (still unsure if i should add it by default on pyjnius or not), here is a way that i'm testing on an app. For any Python thread created, it will use a monkey-patched method that call jnius.detach():


```python
# alternative approach that works when the thread got an exception as well
import threading
try:
    import jnius
except:
    jnius = None

if jnius:
    orig_thread_run = threading.Thread.run
    def thread_check_run(*args, **kwargs):
        try:
            return ret
        finally:
            jnius.detach()
    threading.Thread.run = thread_check_run
```

See my tests here: https://gist.github.com/tito/09c42fb4767721dc323d

**Comment by Cheaterman** [comment]
https://github.com/kivy/pyjnius/commit/98386606262329b29cf74bc894b048ac37dbd6cb#commitcomment-33704483

:-(

FWIW, the `UrlRequest`-like class is below (see lines 148-149) :

`import http.client
import io
import json
import queue
import threading
import time

from kivy.app import App
from kivy.clock import Clock
import requests
import requests_toolbelt.utils.dump


class BTRequest(threading.Thread):
    """Very minimal UrlRequest-like class for HTTP over Bluetooth RFCOMM"""

    def __init__(
        self,
        url,
        on_success=None,
        on_failure=None,
        on_error=None,
        req_body=None,
        req_headers=None,
        chunk_size=8192,
        timeout=None,
        method=None,
    ):
        super().__init__(daemon=True)

        self.url = url
        self.on_success = on_success
        self.on_failure = on_failure
        self.on_error = on_error
        self.req_body = req_body
        self.req_headers = req_headers
        self.chunk_size = chunk_size
        self.timeout = timeout
        self.method = method
        self.event_queue = queue.Queue()
        self.trigger_result = Clock.create_trigger(self.dispatch_result)

        self.start()

    def run(self):
        url = self.url
        if '://' not in url:
            url = f'http://localhost{url}'

        method = self.method
        if method is None:
            method = 'GET' if not self.req_body else 'POST'

        data = bytearray()

        requests_toolbelt.utils.dump._dump_request_data(
            requests.Request(
                method,
                url,
                headers=self.req_headers,
                data=self.req_body,
            ).prepare(),
            requests_toolbelt.utils.dump.PrefixSettings('', ''),
            data
        )

        app = App.get_running_app()

        app.send(bytes(data))

        class BytesIOSocket:
            def makefile(self, *args):
                return io.BytesIO()

        headers_length = None
        content_length = None
        response = http.client.HTTPResponse(BytesIOSocket())
        last_data_time = time.time()

        while True:
            if self.timeout is not None:
                if time.time() - last_data_time >= self.timeout:
                    response.fp.seek(0)
                    self.event_queue.put((
                        'error',
                        None,
                        TimeoutError(
                            'Request timed out after {self.timeout} seconds.'
                        )
                    ))
                    break

            data = app.recv(self.chunk_size)

            if data:
                response.fp.write(data)
                last_data_time = time.time()
            elif not response.fp.tell():
                time.sleep(.01)
                continue

            if headers_length is None:
                response.fp.seek(0)
                response.begin()
                current_position = response.fp.tell()

                if(
                    current_position == response.fp.seek(0, io.SEEK_END)
                    and int(response.getheader('content-length', '-1')) != 0
                ):
                    response.headers = None
                    continue

                else:
                    headers_length = current_position
                    content_length = int(
                        response.getheader('content-length', '0')
                    )

            else:
                current_size = response.fp.tell()

                if current_size - headers_length >= content_length:
                    response.fp.seek(headers_length)
                    result = response.read()

                    if response.getheader(
                        'content-type',
                        'text/plain'
                    ).split(';', 1)[0] == 'application/json':
                        if isinstance(result, bytes):
                            result = result.decode('utf8')

                        try:
                            result = json.loads(result)
                        except json.JSONDecodeError:
                            pass

                    self.event_queue.put(('success', response, result))
                    break

            time.sleep(.01)

        while not self.event_queue.empty():
            self.trigger_result()
            time.sleep(.01)

        # No crash if I call this:
        #super().run()

    def dispatch_result(self, _):
        while True:
            try:
                result, response, data = self.event_queue.get_nowait()
            except queue.Empty:
                return

            if result == 'success':
                status_code = response.status

                if 100 <= status_code <= 299:
                    if self.on_success:
                        self.on_success(self, data)

                elif 400 <= status_code <= 599:
                    if self.on_failure:
                        self.on_failure(self, data)

            elif result == 'error':
                if self.on_error:
                    self.on_error(self, data)
`
